server/src/main/java/org/elasticsearch/persistent/AllocatedPersistentTask.java

```diff
@@ -99,30 +100,49 @@ boolean markAsCancelled() {
     }
 
     public void markAsCompleted() {
-        completeAndNotifyIfNeeded(null);
+        completeAndNotifyIfNeeded(null, null);
     }
 
     public void markAsFailed(Exception e) {
         if (CancelTasksRequest.DEFAULT_REASON.equals(getReasonCancelled())) {
-            completeAndNotifyIfNeeded(null);
+            completeAndNotifyIfNeeded(null, null);
         } else {
-            completeAndNotifyIfNeeded(e);
+            completeAndNotifyIfNeeded(e, null);
         }
     }
 
-    private void completeAndNotifyIfNeeded(@Nullable Exception failure) {
+    /**
+     * Indicates that this persistent task is no longer going to run on the local node.
+     * This will cause the local task to be terminated, and the associated persistent
+     * task to be reassigned by the master node. The persistent task <em>may</em> be
+     * reassigned to the same node unless separate measures have been taken to prevent
+     * this. The task should complete any graceful shutdown actions before calling this
+     * method.
+     * @param localAbortReason Reason for the task being aborted on this node. This
+     *                         will be recorded as the reason for unassignment of the
+     *                         persistent task.
+     */
+    public void markAsLocallyAborted(String localAbortReason) {
+        completeAndNotifyIfNeeded(null, Objects.requireNonNull(localAbortReason));
```

Comment:
I am not sure it's worth adding the complexity of two methods given that this is likely to have quite limited uses in the remaining life of 7.x.

For ML, falling back to `markAsFailed` would always be the wrong thing, and really bad.  It would mean that after using the node shutdown API in a mixed version 7.x cluster a user would have to manually intervene to restart ML jobs that unexpectedly failed during upgrade.  If the person doing the upgrade was different to the person who created the ML jobs and they weren't expecting any ML jobs to fail then those jobs might not run for weeks.

So, what do you think about the idea of having this method log a warning, throw an exception and do nothing if the minimum cluster node version is too old for every node to understand the new way of stopping?

Then it will be up to the ML shutdown code to know that it shouldn't try and do anything graceful or clever in a mixed 7.x cluster containing pre-7.15 nodes.  That will be no worse than the situation everyone is in with current releases.  It just means that ML jobs might need to do a non-trivial amount of rework after starting up on a different node after upgrade.

I suspect no sub-team other than ML will have to deal with this, because nobody else will use it during the remaining lifetime of 7.x.  Or if they do, they could add the second method `markAsLocallyAbortedOrFailed` at the time it's actually useful.

Reply:
That sounds like a good compromise to me. Let us ensure we document the behavior on the method.

Reply:
Now noticing that you already made this change, looks good.